core/src/main/java/software/amazon/awssdk/core/sync/ResponseTransformer.java

```diff
@@ -157,7 +158,11 @@ default boolean needsConnectionLeftOpen() {
      */
     static <ResponseT> ResponseTransformer<ResponseT, ResponseT> toOutputStream(OutputStream outputStream) {
         return (resp, in) -> {
-            IoUtils.copy(in, outputStream);
+            try {
+                IoUtils.copy(in, outputStream);
```

Comment:
Ideally we'd do this at a higher layer so that any response transformer implementation that is passed in gets it's exceptions converted to non-retryable exceptions unless they explicitly throw a retryable exception.

Reply:
Yeah that would be nice indeed. 
Unfortunately `ResponseTransformer` is used all over the place so there isn't a higher layer to put the logic currently. 

I guess we could introduce one within this class, something like:

```java
@FunctionalInterface
@SdkPublicApi
public interface ResponseTransformer<ResponseT, ReturnT> {
    /**
     * Process the response contents.
     *
     * @param response    Unmarshalled POJO response
     * @param inputStream Input stream of streamed data.
     * @return Transformed type.
     * @throws Exception if any error occurs during processing of the response. This will be re-thrown by the SDK, possibly
     *                   wrapped in an {@link SdkClientException}.
     */
    default ReturnT transform(ResponseT response, AbortableInputStream inputStream) throws Exception {
        try {
            return apply(response, inputStream);
        } catch (RetryableException e) {
            throw e;
        } catch (Exception e) {
            throw new NonRetryableException(e);
        }
    }

    ReturnT apply(ResponseT response, AbortableInputStream inputStream) throws Exception;
}
```

I'm not familiar with this codebase so I may be completely off track 🤷‍♂️